exporter/sumologicexporter/exporter.go

```diff
@@ -16,44 +16,113 @@ package sumologicexporter
 
 import (
 	"context"
+	"net/http"
+
+	"go.opentelemetry.io/collector/component/componenterror"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
 
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/consumer/pdata"
-	"go.uber.org/zap"
+)
+
+const (
+	logKey        string = "log"
+	maxBufferSize int    = 100
 )
 
 type sumologicexporter struct {
-	logger *zap.Logger
+	config *Config
+	client *http.Client
+	filter *filtering
 }
 
-var _ (component.LogsExporter) = (*sumologicexporter)(nil)
-
 func newLogsExporter(
-	logger *zap.Logger,
+	cfg *Config,
 ) (component.LogsExporter, error) {
-	se := &sumologicexporter{logger: logger}
-	return se, nil
+	se, err := initExporter(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return exporterhelper.NewLogsExporter(
+		cfg,
+		se.pushLogsData,
+		// Disable exporterhelper Timeout, since we are using a custom mechanism
+		// within exporter itself
+		exporterhelper.WithTimeout(exporterhelper.TimeoutSettings{Timeout: 0}),
+		exporterhelper.WithRetry(cfg.RetrySettings),
+		exporterhelper.WithQueue(cfg.QueueSettings),
+	)
 }
 
-func newMetricsExporter(
-	logger *zap.Logger,
-) (component.MetricsExporter, error) {
-	se := &sumologicexporter{logger: logger}
+func initExporter(cfg *Config) (*sumologicexporter, error) {
+	filter, err := newFiltering(cfg.MetadataAttributes)
+	if err != nil {
+		return nil, err
+	}
+
+	se := &sumologicexporter{
+		config: cfg,
+		client: &http.Client{
+			Timeout: cfg.TimeoutSettings.Timeout,
+		},
+		filter: filter,
+	}
+
 	return se, nil
 }
 
-func (se *sumologicexporter) Start(_ context.Context, _ component.Host) error {
-	return nil
-}
+// pushLogsData groups data with common metadata and send them together to Sumo Logic
+func (se *sumologicexporter) pushLogsData(ctx context.Context, ld pdata.Logs) (droppedTimeSeries int, err error) {
+	var (
+		currentMetadata  FieldsType
+		previousMetadata FieldsType
+		errors           []error
+		sdr              *sender = newSender(se.config, se.client, se.filter)
+	)
 
-func (se *sumologicexporter) Shutdown(context.Context) error {
-	return nil
-}
+	// Iterate over ResourceLogs
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		resource := ld.ResourceLogs().At(i)
 
-func (se *sumologicexporter) ConsumeLogs(_ context.Context, _ pdata.Logs) error {
-	return nil
-}
+		// iterate over InstrumentationLibraryLogs
+		for j := 0; j < resource.InstrumentationLibraryLogs().Len(); j++ {
+			library := resource.InstrumentationLibraryLogs().At(j)
+
+			// iterate over Logs
+			for k := 0; k < library.Logs().Len(); k++ {
+				log := library.Logs().At(k)
+				currentMetadata = sdr.filter.GetMetadata(log.Attributes())
+
+				// If metadata differs from currently buffered, flush the buffer
+				if currentMetadata != previousMetadata && previousMetadata != "" {
+					dropped, err := sdr.sendLogs(previousMetadata)
+					if err != nil {
+						droppedTimeSeries += dropped
+						errors = append(errors, err)
```

Comment:
I am not sure what's going on here and what's the intention.

1) Is this an error really? From what I can tell, in such scenario data is still being processed
2) Wouldn't it be better to group data my common metadata first?

Reply:
ad.2 I believe it should be done by some processor not in the exporter itself. Exporter should just push the data. Also as it is not so safe from memory management point of view to group data in expoerter I would prefer to not do it in first iteration

ad.1 If we fail to send some logs it is error, but we should keep pushing rest of the logs and return total stats and combine errors for all requests